# Side scripts
Additional packages needed when using side scripts (**None** of them are needed to run
Predmoter or side scripts directly connected to Predmoter like ``add_blacklist.py``.):
- sklearn/scikit-learn (used in: ``compute_peak_f1.py``, version used: 1.3.0)
- configobj (used in: ``setup_qsubs.py``)

## Script overview
| Script                 | Description                                                                                                                                                                                                                                                                               |
|:-----------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| add_blacklist.py       | Add ``data/blacklist`` to training/validation and/or test h5 files to mask non-nuclear and scaffold regions, if the assembly doesn't only contain scaffolds/contigs.                                                                                                                      |
| add_metrics.py         | Add averages and/or medians of chosen NGS dataset(s) to the h5 file (**NOT** a prediction h5 file, but one created by Helixer to train, validate or test on).                                                                                                                             |
| compute_peak_f1.py     | Computes the F1, true positives (TP), false positives (FP), and false negatives (FN) between two bed files (e.g., experimental and predicted) containing called peaks (e.g., ATAC-seq peaks) from one species. The peaks are assumed to have been called with MACS3.                      |
| extract_metrics.py     | Extract metrics (mean or median) from a h5 file. The metrics have to have been added with ``add_metrics.py``.                                                                                                                                                                             |
| pearson_coeff_rep.py   | Calculate Pearson correlation for each experimental NGS data replicate combination added in a given h5 file.                                                                                                                                                                              |
| replace_adapters.sh    | When setting up your pipeline to process public NGS data with [RNAsleek](https://github.com/weberlab-hhu/RNAsleek) and ``setup_qsubs.py``, this script replaces the standard ``TruSeq3-PE-2.fa`` adapters to the ``NexteraPE-PE.fa`` adapters, which are common for public ATAC-seq data. |
| setup_qsubs.py         | Setup additional qsub files needed for processing public ATAC/ChIP-seq data when using [RNAsleek](https://github.com/weberlab-hhu/RNAsleek).                                                                                                                                              |
| extract_peak_length.py | Computes either the mean or median peak length of all peaks inside a given MACS3 peak bed file. Unwanted chromosomes/sequence IDs can be excluded via a blacklist file, created with add_blacklist.py, like for compute_peak_f1.                                                          |
| extract_peak_number.py | Computes the number of all peaks inside a given MACS3 peak bed file. Unwanted chromosomes/sequence IDs can be excluded via a blacklist file, created with add_blacklist.py, like for compute_peak_f1.                                                                                     |
